pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -0,0 +1,993 @@
+// each region.
+
+func (lbaas *LbaasV2) EnsureLoadBalancer(apiService *api.Service, hosts []string, annotations map[string]string) (*api.LoadBalancerStatus, error) {
+	glog.V(4).Infof("EnsureLoadBalancer(%v, %v, %v, %v, %v, %v)", apiService.Namespace, apiService.Name, apiService.Spec.LoadBalancerIP, apiService.Spec.Ports, hosts, annotations)
+
+	ports := apiService.Spec.Ports
+	if len(ports) > 1 {
+		return nil, fmt.Errorf("multiple ports are not yet supported in openstack load balancers")
+	} else if len(ports) == 0 {
+		return nil, fmt.Errorf("no ports provided to openstack load balancer")
+	}
+
+	// The service controller verified all the protocols match on the ports, just check and use the first one
+	// TODO: Convert all error messages to use an event recorder
+	if ports[0].Protocol != api.ProtocolTCP {
+		return nil, fmt.Errorf("Only TCP LoadBalancer is supported for openstack load balancers")
+	}
+
+	affinity := api.ServiceAffinityNone //apiService.Spec.SessionAffinity
+	var persistence *v2_pools.SessionPersistence
+	switch affinity {
+	case api.ServiceAffinityNone:
+		persistence = nil
+	case api.ServiceAffinityClientIP:
+		persistence = &v2_pools.SessionPersistence{Type: "SOURCE_IP"}
+	default:
+		return nil, fmt.Errorf("unsupported load balancer affinity: %v", affinity)
+	}
+
+	sourceRanges, err := service.GetLoadBalancerSourceRanges(annotations)
+	if err != nil {
+		return nil, err
+	}
+
+	if !service.IsAllowAll(sourceRanges) {
+		return nil, fmt.Errorf("Source range restrictions are not supported for openstack load balancers")
+	}
+
+	glog.V(2).Infof("Checking if openstack load balancer already exists: %s", cloudprovider.GetLoadBalancerName(apiService))
+	_, exists, err := lbaas.GetLoadBalancer(apiService)
+	if err != nil {
+		return nil, fmt.Errorf("error checking if openstack load balancer already exists: %v", err)
+	}
+
+	// TODO: Implement a more efficient update strategy for common changes than delete & create
+	// In particular, if we implement hosts update, we can get rid of UpdateHosts
+	if exists {
+		err := lbaas.EnsureLoadBalancerDeleted(apiService)
+		if err != nil {
+			return nil, fmt.Errorf("error deleting existing openstack load balancer: %v", err)
+		}
+	}
+
+	lbmethod := v2_pools.LBMethod(lbaas.opts.LBMethod)
+	if lbmethod == "" {
+		lbmethod = v2_pools.LBMethodRoundRobin
+	}
+	name := cloudprovider.GetLoadBalancerName(apiService)
+
+	createOpts := loadbalancers.CreateOpts{
+		Name:         name,
+		Description:  fmt.Sprintf("Kubernetes external service %s", name),
+		VipSubnetID:  lbaas.opts.SubnetId,
+	}
+
+	loadBalancerIP := apiService.Spec.LoadBalancerIP
+	if loadBalancerIP != "" {
+		createOpts.VipAddress = loadBalancerIP
+	}
+
+	loadbalancer, err := loadbalancers.Create(lbaas.network, createOpts).Extract()
+	if err != nil {
+		// cleanup what was created so far
+		_ = lbaas.EnsureLoadBalancerDeleted(apiService)
+		return nil, err
+	}
+
+	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
```

Comment:
yeah, unfortunately anytime a change is made to a component, the respective load balancer changes state to "... pending".  During this time, any changes made will return with failure bc the load balancer state is not ACTIVE